iceoryx_utils/include/iceoryx_utils/cxx/forward_list.hpp

```diff
@@ -0,0 +1,375 @@
+    /// @param[in] ur_data universal reference perfectly forwarded to client class
+    /// @return successful insertion (true), otherwise no element could be added to list (e.g. full -> false)
+    bool push_front(T&& ur_data) noexcept;
+
+    /// @brief remove the first element from the begining of the list
+    ///         element destructor will be invoked
+    /// @return successful removal (true), otherwise no element could be taken from list (e.g. empty -> false)
+    bool pop_front() noexcept;
+
+    // possible future function extension
+    // void swap(list & rhs) noexcept; // PRQA S 1051
+
+    /// @brief remove all elements from the list, list will be empty
+    ///         element destructors will be invoked
+    void clear();
+
+    /// @brief remove next element from linked iterator position
+    ///         element destructors will be invoked
+    ///         recursive calls to erase_after only delete each 2nd element
+    /// @param[in] beforeToBeErasedIter iterator linking the element before the to-be-removed element
+    /// @return an (non-const_) iterator to the element after the removed element,
+    ///         returns end() element when reached end of list
+    iterator erase_after(const_iterator beforeToBeErasedIter) noexcept;
+
+    /// @brief construct element inplace at begining of list
+    /// @param[in] args T-typed construction parameters (initializer list)
+    template <typename... ConstructorArgs>
+    /// @return successful insertion (true), otherwise no element could be added to list (e.g. full -> false)
+    bool emplace_front(ConstructorArgs&&... args) noexcept;
+
+    /// @brief construct element inplace at begining of list
+    /// @param[in] args T-typed construction parameters (initializer list)
+    /// @param[in] afterToBeEmplacedIter position in list to (construct)insert after
+    /// @return iterator to the newly added element
+    template <typename... ConstructorArgs>
+    iterator emplace_after(const_iterator afterToBeEmplacedIter, ConstructorArgs&&... args) noexcept;
+
+    /// @brief insert element after iterator position
+    /// @param[in] citer iterator with the position to insert after
+    /// @param[in] r_data reference to element to add
+    /// @return iterator to the newly added element
+    iterator insert_after(const_iterator citer, const T& r_data) noexcept;
+
+    /// @brief construct element inplace at begining of list
+    /// @param[in] citer iterator with the position to insert after
+    /// @param[in] ur_data universal reference perfectly forwarded to client class
+    /// @return iterator to the newly added element
+    iterator insert_after(const_iterator citer, T&& ur_data) noexcept;
+
+  private:
+    struct ListNodeType
+    {
+        T data;
+        sizeType nextIdx;
+    };
+
+    void init() noexcept;
+    ListNodeType* getNodePtrFromIdx(sizeType) const noexcept;
+    ListNodeType* getHeadFreeEl() const noexcept;
```

Comment:
Please return `const ListNodeType *` since this method is const and we do not want to change the internal state.

What does `getHeadFreeEl` mean? Please do not use abbreviation which are not commonly known.